pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,118 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pod
+
+import (
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	pdutil "github.com/pingcap/tidb-operator/pkg/manager/member"
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+	operatorUtils "github.com/pingcap/tidb-operator/pkg/util"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	"k8s.io/api/admission/v1beta1"
+	apps "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (pc *PodAdmissionControl) admitDeletePdPods(pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+	ordinal, err := operatorUtils.GetOrdinalFromPodName(pod.Name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isMember, err := IsPodInPdMembers(tc, pod, pdClient)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isOutOfOrdinal, err := operatorUtils.IsPodOutOfOrdinal(pod, *ownerStatefulSet.Spec.Replicas)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	isUpgrading := IsStatefulSetUpgrading(ownerStatefulSet)
+
+	klog.Infof("receive delete pd pod[%s/%s] of tc[%s/%s],isMember=%v,isOutOfOrdinal=%v,isUpgrading=%v", namespace, name, namespace, tcName, isMember, isOutOfOrdinal, isUpgrading)
+
+	// each pd instance must be member deleted from tc-cluster before deleting its pod.
+	if !isMember {
+		// when pd scale in,we should delete member first and finally edit its pvc and admit to delete pod.
+		// If we edit pvc first and finally delete pd member during scale in,
+		// it can be error that pd scale in From 4 to 3,the pvc was edited successfully and
+		// the pd member was fail to deleted. If the pd scale was recover to 4 at that time,
+		// it would be existed an pd-3 instance with its deferDeleting label Annotations PVC.
+		// And the pvc can be deleted during upgrading if we use create pod webhook in future.
+		if isOutOfOrdinal {
+			err = addDeferDeletingToPVC(pc, tc, ownerStatefulSet.Name, namespace, ordinal)
+			if err != nil {
+				klog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
+				return util.ARFail(err)
+			}
+		}
+		klog.Infof("pd pod[%s/%s] is not member of tc[%s/%s],admit to delete", namespace, name, namespace, tcName)
+		return util.ARSuccess()
+	}
+
+	if isOutOfOrdinal {
+		err = pdClient.DeleteMember(name)
+		if err != nil {
+			return util.ARFail(err)
+		}
+		klog.Infof("tc[%s/%s]'s pd[%s/%s] is being deleted from pd-cluster,refuse to delete it.", namespace, tcName, namespace, name)
+		return &v1beta1.AdmissionResponse{
+			Allowed: false,
+		}
+	}
+
+	if isUpgrading {
```

Comment:
Just make sure the PD service is available, no need to care about the ordinals.

Reply:
ditto.

Reply:
> Just make sure the PD service is available, no need to care about the ordinals.

ditto.